src/burnchains/bitcoin/spv.rs

```diff
@@ -725,56 +740,73 @@ impl SpvClient {
+            0x0000000000000000,
+            0x00000000ffff0000,
+        ]);
+        let max_target_bits = BlockHeader::compact_target_from_u256(&max_target);
+
+        let parent_header = if headers_in_range.len() > 0 {
+            headers_in_range[0]
         } else {
-            Uint256([
-                0x0000000000000000,
-                0x0000000000000000,
-                0x0000000000000000,
-                0x00000000ffff0000,
-            ])
+            match self.read_block_header(current_header_height - 1)? {
+                Some(res) => res.header,
+                None => return Ok(None),
+            }
         };
 
-        if interval == 0 {
-            panic!(
-                "Invalid argument: interval must be positive (got {})",
-                interval
-            );
+        if current_header_height % BLOCK_DIFFICULTY_CHUNK_SIZE != 0
+            && self.network_id == BitcoinNetworkType::Testnet
+        {
+            // In Testnet mode, if the new block's timestamp is more than 2* 10 minutes
+            // then allow mining of a min-difficulty block.
+            if current_header.time > parent_header.time + 10 * 60 * 2 {
+                return Ok(Some((max_target_bits, max_target)));
+            }
+
+            // Otherwise return the last non-special-min-difficulty-rules-block
+            for ancestor in headers_in_range {
+                if ancestor.bits != max_target_bits {
+                    let target = ancestor.target();
+                    return Ok(Some((ancestor.bits, target)));
+                }
+            }
         }
 
         let first_header =
             match self.read_block_header((interval - 1) * BLOCK_DIFFICULTY_CHUNK_SIZE)? {
-                Some(header) => header,
-                None => {
-                    // haven't got here yet
-                    return Ok(None);
-                }
+                Some(res) => res.header,
+                None => return Ok(None),
             };
 
         let last_header =
             match self.read_block_header(interval * BLOCK_DIFFICULTY_CHUNK_SIZE - 1)? {
-                Some(header) => header,
-                None => {
-                    // whatever the current header is
-                    return Ok(None);
-                }
+                Some(res) => res.header,
+                None => return Ok(None),
             };
 
         // find actual timespan as being clamped between +/- 4x of the target timespan
-        let measured_timespan = (last_header.header.time - first_header.header.time) as u64;
+        let mut actual_timespan = (last_header.time - first_header.time) as u64;
         let target_timespan = BLOCK_DIFFICULTY_INTERVAL as u64;
-        let timespan_highpass = cmp::max(measured_timespan, target_timespan / 4);
-        let filtered_timespan = cmp::min(timespan_highpass, target_timespan * 4);
+        if actual_timespan < (target_timespan / 4) {
+            actual_timespan = target_timespan / 4;
```

Comment:
Please revert this change between lines 793 and 809.  The original code better represents what's actually going on inside bitcoind.

Reply:
Can you share with me a link of the code in bitcoin that looks like the previous version? It probably has been updated since then, I've actually been looking at the code of Bitcoin, and this version is, factually, easier to compare to the original version:

```
unsigned int CalculateNextWorkRequired(const CBlockIndex* pindexLast, int64_t nFirstBlockTime, const Consensus::Params& params)
{
    if (params.fPowNoRetargeting)
        return pindexLast->nBits;


    // Limit adjustment step
    int64_t nActualTimespan = pindexLast->GetBlockTime() - nFirstBlockTime;
    if (nActualTimespan < params.nPowTargetTimespan/4)
        nActualTimespan = params.nPowTargetTimespan/4;
    if (nActualTimespan > params.nPowTargetTimespan*4)
        nActualTimespan = params.nPowTargetTimespan*4;


    // Retarget
    const arith_uint256 bnPowLimit = UintToArith256(params.powLimit);
    arith_uint256 bnNew;
    bnNew.SetCompact(pindexLast->nBits);
    bnNew *= nActualTimespan;
    bnNew /= params.nPowTargetTimespan;
   ...
```

(https://github.com/bitcoin/bitcoin/blob/78dae8caccd82cfbfd76557f1fb7d7557c7b5edb/src/pow.cpp#L49-L70)

Reply:
I stand corrected.  If that's what's currently in bitcoind, then that's what we should use :+1: